include/util/opening_hours.hpp

```diff
@@ -0,0 +1,645 @@
+            using boost::gregorian::date_duration;
+
+            const auto year = time.tm_year + 1900;
+            const auto month = time.tm_mon + 1;
+
+            date date_current(year, month, time.tm_mday);
+            date date_from(boost::gregorian::min_date_time);
+            date date_to(boost::gregorian::max_date_time);
+
+            if (from.IsValid())
+            {
+                date_from = (from.day == 0) ? date(from.year == 0 ? year : from.year,
+                                                   from.month == 0 ? month : from.month,
+                                                   1)
+                                            : date(from.year == 0 ? year : from.year,
+                                                   from.month == 0 ? month : from.month,
+                                                   from.day);
+            }
+            if (to.IsValid())
+            {
+                date_to = date(to.year == 0 ? (from.year == 0 ? year : from.year) : to.year,
+                               to.month == 0 ? (from.month == 0 ? month : from.month) : to.month,
+                               1);
+                date_to = (to.day == 0) ? date_to.end_of_month()
+                                        : date(date_to.year(), date_to.month(), to.day);
+            }
+            else if (to == Monthday())
+            {
+                date_to = date_from;
+            }
+
+            if (!use_curr_day)
+                date_from += date_duration(1);
+            if (use_next_day && date_to != date(boost::gregorian::max_date_time))
+                date_to += date_duration(1);
+
+            return date_from <= date_current && date_current <= date_to;
+        }
+    };
+
+    OpeningHours() : modifier(open) {}
+
+    bool IsInRange(const struct tm &time) const
+    {
+        bool use_curr_day = true;  // the first matching time uses the current day
+        bool use_next_day = false; // the first matching time uses the next day
+        return (times.empty() ||
+                std::any_of(times.begin(),
+                            times.end(),
+                            [&time, &use_curr_day, &use_next_day](const auto &x) {
+                                return x.IsInRange(time, use_curr_day, use_next_day);
+                            })) &&
+               (weekdays.empty() ||
+                std::any_of(weekdays.begin(),
+                            weekdays.end(),
+                            [&time, use_curr_day, use_next_day](const auto &x) {
+                                return x.IsInRange(time, use_curr_day, use_next_day);
+                            })) &&
+               (monthdays.empty() ||
+                std::any_of(monthdays.begin(),
+                            monthdays.end(),
+                            [&time, use_curr_day, use_next_day](const auto &x) {
+                                return x.IsInRange(time, use_curr_day, use_next_day);
+                            }));
```

Comment:
I can add human-readable comments, because it uses logical shortcuts, otherwise it will be like
```
if (!bla)
{
   return false;
}
```